### ISSUE TITLE
String matching using wildcards

### ISSUE THREAD

**Issue by afs** [w3c/sparql-12#85]
Regular expressions can be complex.  Strings with wildcards are simpler.

## Proposed solution

Provide string matching using wildcards as an additional, alternative to regular
expressions by adding a new function. The string is anchored.

Examples:

    MATCH(?string, "abc*")

    MATCH(?string, "*abc*", "i") # Case insentive.

    MATCH(?string, "a?c")

## Previous work

[Glob patterns](https://en.wikipedia.org/wiki/Glob_(programming))
[SQL LIKE](https://en.wikipedia.org/wiki/Where_(SQL)#LIKE)
[Lucene wildcard searches](https://lucene.apache.org/core/2_9_4/queryparsersyntax.html#Wildcard%20Searches)

## Considerations for backward compatibility

None.

**Comment by lisp** [comment]
why?

**Comment by cygri** [comment]
If this were added, it should be a different name. When explaining SPARQL, one constantly has to talk about matching—and it usually means matching graph patterns against triples. Having a function called “match” that uses the word in a completely different sense does not help.

Some possible other names:
```sparql
FILTER wildcard(?title, "*sparql*", "i")
FILTER like(?title, "*sparql*", "i")
```
This could also be combined with #34:
```sparql
?doc :title ~"*sparql*"i.
```

**Comment by dbooth-boston** [comment]
It is called  `glob` in [several other languages](https://en.wikipedia.org/wiki/Glob_(programming)).

**Comment by lisp** [comment]
if the goal is succinctness, it makes sense to go all the way to something like

> ```sparql
> ?doc :title ~"*sparql*"i.
> ```

**Comment by VladimirAlexiev** [comment]
Shex has a similar construct called Stem.
It works for strings, IRIs (also prefixed) and lang tags

**Comment by afs** [comment]
Shex Stem is [fn:starts-with](https://www.w3.org/TR/xpath-functions-31/#func-starts-with) / `STRSTARTS`.

**Comment by afs** [comment]
I agree "match" is already used for graph patterns,. It is also valuable a as a keyword.

`LIKE` is good depending on the SQL implications (SQL uses `_` and `%` for what is commonly `*` and `?` in shells and filename matching); SQL LIKE also has character classes and negated character classes.

Filename matching with glob matching, where `*` means any character except the component separator, and some systems (e.g. git) add `**` to mean "filename, any depth".

Possibilities:
* `STRMATCH`
* `LIKE`
* `GLOB`
* `WILDCARD`
* . . .

The other choice s what matching language.

SQL LIKE can be rewritten to a regex expression and there are code examples for that online. 

In Java, there a few open source direct implementations with `*` and `?`, but not `[ ] ` character classes. (The JDK supports glob on filenames, not directly for strings).

**Comment by ktk** [comment]
I think being close to SQL is not a bad idea so I like the `LIKE` idea. If I would not know about that I would go for `STRMATCH` as it resembles other functions in SPARQL but then again it might add to the confusion. `LIKE` is unique in that sense.

And while I like the `GLOB` idea I have to agree that I mainly know it for file matching.
